MediaPlayerPrivateRemote should derive privately from IPC::MessageReceiver

youennf opened the pull request:
#### c54171089372d22c1838e8ce92e1ec9dc638d3d6
<pre>
MediaPlayerPrivateRemote should derive privately from IPC::MessageReceiver
<a href="https://rdar.apple.com/130574962">rdar://130574962</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275896">https://bugs.webkit.org/show_bug.cgi?id=275896</a>

Reviewed by Eric Carlson.

This clarifies which base class weak pointer to take for MediaPlayerPrivateRemote.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load):
(WebKit::MediaPlayerPrivateRemote::performTaskAtTime):
(WebKit::MediaPlayerPrivateRemote::requestHostingContextID):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::createRemoteMediaPlayer):
(WebKit::RemoteMediaPlayerManager::deleteRemoteMediaPlayer):
(WebKit::RemoteMediaPlayerManager::findRemotePlayerId):
(WebKit::RemoteMediaPlayerManager::gpuProcessConnectionDidClose):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/VideoLayerRemoteCocoa.mm:
(-[WKVideoLayerRemote mediaPlayerPrivateRemote]):
(-[WKVideoLayerRemote resizePreservingGravity]):
(-[WKVideoLayerRemote layoutSublayers]):
(-[WKVideoLayerRemote resolveBounds]):

Canonical link: <a href="https://commits.webkit.org/280415@main">https://commits.webkit.org/280415@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6eeefc3d59b833eb3d551a13b5342cf6c370f523

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56580 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9053 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60191 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7021 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43530 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7212 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45828 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4908 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58610 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33756 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48828 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26689 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30537 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6024 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6432 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61874 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/490 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6536 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53086 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/492 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48888 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53050 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12519 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/421 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31735 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32821 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33905 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32567 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -270,8 +270,9 @@ void MediaPlayerPrivateRemote::load(const URL& url, const ContentType& contentTy
-    connection().sendWithAsyncReply(Messages::RemoteMediaPlayerProxy::Load(url, WTFMove(sandboxExtensionHandle), contentType, keySystem, m_player.get()->requiresRemotePlayback()), [weakThis = WeakPtr { *this }, this](auto&& configuration) {
-        if (!weakThis)
+    connection().sendWithAsyncReply(Messages::RemoteMediaPlayerProxy::Load(url, WTFMove(sandboxExtensionHandle), contentType, keySystem, m_player.get()->requiresRemotePlayback()), [weakThis = ThreadSafeWeakPtr { *this }, this](auto&& configuration) {
+        auto protectedThis = weakThis.get();
```
Isn't the new suggested pattern `RefPtr protectedThis = ...`?

youennf replied:
Ah right, will fix it.

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -1007,8 +1008,9 @@ void MediaPlayerPrivateRemote::load(const URL& url, const ContentType& contentTy
-        connection().sendWithAsyncReply(Messages::RemoteMediaPlayerProxy::LoadMediaSource(url, contentType, DeprecatedGlobalSettings::webMParserEnabled(), identifier), [weakThis = WeakPtr { *this }, this](RemoteMediaPlayerConfiguration&& configuration) {
-            if (!weakThis)
+        connection().sendWithAsyncReply(Messages::RemoteMediaPlayerProxy::LoadMediaSource(url, contentType, DeprecatedGlobalSettings::webMParserEnabled(), identifier), [weakThis = ThreadSafeWeakPtr { *this }, this](RemoteMediaPlayerConfiguration&& configuration) {
+            auto protectedThis = weakThis.get();
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -1023,8 +1025,9 @@ void MediaPlayerPrivateRemote::load(const URL& url, const ContentType& contentTy
-    callOnMainRunLoop([weakThis = WeakPtr { *this }, this] {
-        if (!weakThis)
+    callOnMainRunLoop([weakThis = ThreadSafeWeakPtr { *this }, this] {
+        auto protectedThis = weakThis.get();
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -1040,8 +1043,9 @@ void MediaPlayerPrivateRemote::load(const URL& url, const ContentType& contentTy
-    callOnMainRunLoop([weakThis = WeakPtr { *this }, this] {
-        if (!weakThis)
+    callOnMainRunLoop([weakThis = ThreadSafeWeakPtr { *this }, this] {
+        auto protectedThis = weakThis.get();
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -1584,9 +1588,11 @@ void MediaPlayerPrivateRemote::setPreferredDynamicRangeMode(WebCore::DynamicRang
-    auto asyncReplyHandler = [weakThis = WeakPtr { *this }, task = WTFMove(task)](std::optional<MediaTime> currentTime) mutable {
-        if (!weakThis || !currentTime)
+    auto asyncReplyHandler = [weakThis = ThreadSafeWeakPtr { *this }, task = WTFMove(task)](std::optional<MediaTime> currentTime) mutable {
+        auto protectedThis = weakThis.get();
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -1708,9 +1714,9 @@ void MediaPlayerPrivateRemote::requestHostingContextID(LayerHostingContextIDCall
-        if (weakThis)
-            weakThis->setLayerHostingContextID(contextID);
+    connection().sendWithAsyncReply(Messages::RemoteMediaPlayerProxy::RequestHostingContextID(), [weakThis = ThreadSafeWeakPtr { *this }] (auto contextID) {
+        if (auto protectedThis = weakThis.get())
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
```diff
@@ -275,11 +275,10 @@ void RemoteMediaPlayerManager::gpuProcessConnectionDidClose(GPUProcessConnection
-            player->player()->reloadAndResumePlaybackIfNeeded();
-            ASSERT_WITH_MESSAGE(!player, "reloadAndResumePlaybackIfNeeded should destroy this player and construct a new one");
+    for (auto& player : copyToVector(m_players.values())) {
+        if (auto protectedPlayer = player.get()) {
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/cocoa/VideoLayerRemoteCocoa.mm:
```diff
@@ -103,7 +103,7 @@ - (void)setVideoGravity:(WebCore::MediaPlayerEnums::VideoGravity)videoGravity
 - (bool)resizePreservingGravity
 {
-    auto* player = self.mediaPlayerPrivateRemote;
+    RefPtr<WebKit::MediaPlayerPrivateRemote> player = self.mediaPlayerPrivateRemote;
```
Will `RefPtr player = ...` work here?

youennf replied:
Not on my machine, it says that it is not able to infer the template type, I am not sure why. Maybe ObjC?

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/cocoa/VideoLayerRemoteCocoa.mm:
```diff
@@ -195,7 +195,7 @@ - (void)resolveBounds
     if (!CGRectEqualToRect(self.videoLayerFrame, self.bounds)) {
         self.videoLayerFrame = self.bounds;
-        if (auto* mediaPlayerPrivateRemote = self.mediaPlayerPrivateRemote) {
+        if (RefPtr<WebKit::MediaPlayerPrivateRemote> mediaPlayerPrivateRemote = self.mediaPlayerPrivateRemote) {
```
Ditto

youennf replied:
Ditto

youennf committed:
MediaPlayerPrivateRemote should derive privately from IPC::MessageReceiver
rdar://130574962
https://bugs.webkit.org/show_bug.cgi?id=275896

Reviewed by Eric Carlson.

This clarifies which base class weak pointer to take for MediaPlayerPrivateRemote.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load):
(WebKit::MediaPlayerPrivateRemote::performTaskAtTime):
(WebKit::MediaPlayerPrivateRemote::requestHostingContextID):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::createRemoteMediaPlayer):
(WebKit::RemoteMediaPlayerManager::deleteRemoteMediaPlayer):
(WebKit::RemoteMediaPlayerManager::findRemotePlayerId):
(WebKit::RemoteMediaPlayerManager::gpuProcessConnectionDidClose):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/VideoLayerRemoteCocoa.mm:
(-[WKVideoLayerRemote mediaPlayerPrivateRemote]):
(-[WKVideoLayerRemote resizePreservingGravity]):
(-[WKVideoLayerRemote layoutSublayers]):
(-[WKVideoLayerRemote resolveBounds]):

Canonical link: https://commits.webkit.org/280415@main
